### PR TITLE
Use unique openings

### PR DESCRIPTION
**Pull request by vondele**
each worker uses a unique subset of the opening book, which is randomized with a consistent seed on all workers.

fixes https://github.com/glinscott/fishtest/issues/478

Needs a recent cutechess (1.2).

### CODE REVIEW

**Comment by vondele** [comment]
I have turned the case of not .epd and .pgn opening books into an assert. I think this shouldn't be possible right now.

**Comment by vondele** [comment]
first testing with just my worker connected seems to work well (all game pairs had unique openings). Would need some updated workers (+updated cutechess) connected to see if this also works well if cutechess runs on different OS.... 

If the same srand, doesn't lead to the same random ordering on openings, we might not fully eliminate duplicates (but improve it).

**Comment by ppigazzini** [comment]
@vondele DEV and workers updated, using cute120

**Comment by vondele** [comment]
All unique fens for 5f523f2bdeb31bb53d26c911

for i in `seq 0 299`; do wget https://dfts-0.pigazzini.it/api/pgn/5f523f2bdeb31bb53d26c911-$i.pgn ; done ; 
grep FEN *pgn | cut -d: -f2 | sort | uniq -c | awk '{print $1}' | sort | uniq -c

I'll run it a second time with the larger pool of workers (this one had already 2).

**Comment by ppigazzini** [comment]
@vondele I use this scripts to manage a test workers fleet on my 32 thread VPS

```
#!/bin/bash
# start_test_fleet.sh
for i in {0..5}; do
  sleep $((${i}*15)) && echo ${i} && nohup bash start_test_worker.sh ${i} &>/dev/null &
done
```

```
#!/bin/bash
# stop_test_fleet.sh
touch ${HOME}/_git/__test_folder{0..5}/fishtest/worker/fish.exit
```

```
#!/bin/bash
# start_test_worker.sh
dir_num=${1:-'00'}
usr_pwd=${2:-'user00'}
test_folder=${HOME}/_git/__test_folder${dir_num}
virtual_env=${test_folder}/fishtest/worker/env

rm -rf ${test_folder}
mkdir -p ${test_folder}
cd ${test_folder}
git clone --single-branch --branch master https://github.com/glinscott/fishtest.git
cd ${test_folder}/fishtest
git config user.email "you@example.com"
git config user.name "your_name"

# add here the upstream branch to be tested
git remote add new-upstream https://github.com/ppigazzini/fishtest
git pull --no-edit new-upstream TEST_ppigazzini_books

# add here the PRs to be tested
git pull --no-edit origin pull/783/head
git pull --no-edit origin pull/785/head

cd "${test_folder}/"fishtest/worker
python3 -m venv ${virtual_env}
${virtual_env}/bin/python3 -m pip install --upgrade pip setuptools wheel
${virtual_env}/bin/python3 -m pip install requests
rm -rf requests

${virtual_env}/bin/python3 worker.py --host dfts-0.pigazzini.it ${usr_pwd} ${usr_pwd} --concurrency 5
```

**Comment by vondele** [comment]
again all game pairs from a unique fen. So, from my point of view this PR ready. Might benefit from a second pair of eyes.

Does does need a merge of the new cutechess in the books repo. @ppigazzini what do you think? It presumably is bug free for our purpose (i.e. non-variants), so I would go ahead if you think that's the thing to do.

**Comment by vondele** [comment]
@ppigazzini is your test worker fleet including the windows worker, or is that standalone.

**Comment by ppigazzini** [comment]
@vondele just started 2 msys2 workers and 2 wsl workers.

According @ianfab that bug should hit only a couple of variants, see https://github.com/ianfab/FishCooking/pull/11#issuecomment-687138341

**Comment by vondele** [comment]
also works fine with your msys2 and wsl workers. 

So I propose I'll update the books repo with the new cutechess binaries. I recall correctly that if we do a worker version update, new cutechess binaries are downloaded by the workers as well, right ?

**Comment by ppigazzini** [comment]
I'm wondering if it's acceptable to have some workers with an old version cutechess custom build.
Otherwise we should add a cutechess version check to be sure that everybody is using the correct version.

**Comment by vondele** [comment]
it is certainly not ideal to have old cutechess versions around. If multiple workers would do this, this would lead to game duplication.

note that we don't build the release version, so that would be hard to get. Just looking at the version number would be sufficient for now.

**Comment by vondele** [comment]
I have added the version check. Works with my worker, and I verified the same check should work with the windows version of cutechess-cli.

File: worker/games.py [hunk 1]
```diff
@@ -89,6 +89,35 @@ def required_net(engine):
...
+           minor = int(m.group(2))
+           patch = int(m.group(3))
+
+    p.wait()
+    p.stdout.close()
+
+    if p.returncode != 0:
+        raise Exception("Failed to find cutechess version info" % (p.returncode))
+
+    if not (major > 1 or (major == 1 and minor >= 2)):
```
Comment (ppigazzini) [inline worker/games.py:118]: `if major < 1 or (major == 1 and minor < 2): `

**Comment by ppigazzini** [comment]
Tested OK with cutechess-cli 1.1.0 (raise) and 1.2.0 (works fine) on linux, msys2, wsl.

**Comment by vondele** [comment]
OK, squashed with your review suggestion.

**Review by ppigazzini — APPROVED** [review]

**Comment by ppigazzini** [comment]
PROD updated, thank you @vondele :)
@noobpwnftw please update the fleet at your convenience :)

**Comment by ppigazzini** [comment]
@noobpwnftw please remember to download the updated cutechess-cli v 1.2.0 from https://github.com/official-stockfish/books

**Comment by noobpwnftw** [comment]
Done @ppigazzini

**Pull request merged**